bokehjs/src/coffee/renderer/glyph/patches.coffee

```diff
@@ -85,24 +129,48 @@ class PatchesView extends Glyph.View
     hits = []
     for i in [0...candidates.length]
       idx = candidates[i]
-      if hittest.point_in_poly(sx, sy, @sxs[idx], @sys[idx])
-        hits.push(idx)
+      sxs = @sxss[idx]
+      sys = @syss[idx]
+      for j in [0...sxs.length]
+        if hittest.point_in_poly(sx, sy, sxs[j], sys[j])
+          hits.push(idx)
 
     result = hittest.create_hit_test_result()
     result['1d'].indices = hits
     return result
 
-  scx: (i) ->
-    sum = 0
-    for sx in @sxs[i]
-      sum += sx
-    return sum / @sxs[i].length
-
-  scy: (i) ->
-    sum = 0
-    for sy in @sys[i]
-      sum += sy
-    return sum / @sys[i].length
+  _get_snap_coord: (array) ->
+      sum = 0
+      for s in array
+        sum += s
+      return sum / array.length
+
+  scx: (i, sx, sy) ->
+    if @sxss[i].length is 1
+      # We don't have discontinuous objects so we're ok
+      return @_get_snap_coord(@sxs[i])
+    else
+      # We have discontinuous objects, so we need to find which 
+      # one we're in, we can use point_in_poly again
+      sxs = @sxss[i]
+      sys = @syss[i]
+      for j in [0...sxs.length]
+        if hittest.point_in_poly(sx, sy, sxs[j], sys[j])
+          return @_get_snap_coord(sxs[j])
+    return null
+
+  scy: (i, sx, sy) ->
+    if @syss[i].length is 1
+      # We don't have discontinuous objects so we're ok
+      return @_get_snap_coord(@sys[i])
+    else
+      # We have discontinuous objects, so we need to find which 
+      # one we're in, we can use point_in_poly again
+      sxs = @sxss[i]
+      sys = @syss[i]
+      for j in [0...sxs.length]
+        if hittest.point_in_poly(sx, sy, sxs[j], sys[j])
+          return @_get_snap_coord(sys[j])
```

Comment:
We should eventually refactor this to avoid duplication, but for now it is enough...